bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClientFactory.java

```diff
@@ -75,15 +83,22 @@ public BigQueryClientFactory(
     this.credentials = bigQueryCredentialsSupplier.getCredentials();
     this.headerProvider = headerProvider;
     this.bqConfig = bqConfig;
+    this.audienceForIdentityToken = null;
+    readSessionToIdentityTokenCache =
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(identityTokenTtlInMillis, TimeUnit.MILLISECONDS)
+            .build(CacheLoader.from(IdentityTokenSupplier::getIdentityToken));
   }
 
   public void setAudienceForIdentityToken(String audienceForIdentityToken) {
-    identityTokenSupplier = new IdentityTokenSupplier(audienceForIdentityToken);
+    this.audienceForIdentityToken = audienceForIdentityToken;
   }
 
   public BigQueryReadClient getBigQueryReadClient() {
     synchronized (readClientMap) {
-      if (!readClientMap.containsKey(this)) {
+      if (!readClientMap.containsKey(this)
+          || (audienceForIdentityToken != null
+              && readSessionToIdentityTokenCache.getIfPresent(audienceForIdentityToken) == null)) {
```

Comment:
If we are unable to get identity token for any reason, then this renders the readClientMap useless.
It will cause new read clients getting created each time which will persist in memory due to the shutdown hook.

Is it possible to add the identity token to an existing read client?

Reply:
That's the idea. The current work is keep a single HeaderProvider that will refresh the token.